src/persist-client/src/fetch.rs

```diff
@@ -0,0 +1,518 @@
+                .collect::<Vec<_>>(),
+        )
+    }
+    Description::new(
+        decode_antichain(desc.lower()),
+        decode_antichain(desc.upper()),
+        decode_antichain(desc.since()),
+    )
+}
+
+/// Propagates metadata from readers alongside a `HollowBatch` to apply the
+/// desired semantics.
+#[derive(Debug, Clone)]
+pub(crate) enum LeasedBatchMetadata<T> {
+    /// Apply snapshot-style semantics to the fetched batch.
+    Snapshot {
+        /// Return all values with time leq `as_of`.
+        as_of: Antichain<T>,
+    },
+    /// Apply listen-style semantics to the fetched batch.
+    Listen {
+        /// Return all values with time in advance of `as_of`.
+        as_of: Antichain<T>,
+        /// Return all values with time leq `until`.
+        until: Antichain<T>,
+    },
+}
+
+/// A token representing one read batch.
+///
+/// It is tradeable via [`crate::fetch::fetch_batch`] for the resulting data
+/// stored in the batch.
+///
+/// # Exchange
+///
+/// You can exchange `LeasedBatch`:
+/// - If `leased_seqno.is_none()`
+/// - By converting it to [`SerdeLeasedBatch`] through
+///   `Self::get_droppable_batch`. [`SerdeLeasedBatch`] is exchangeable,
+///   including over the network.
+///
+/// # Panics
+/// `LeasedBatch` panics when dropped unless a very strict set of invariants are
+/// held:
+///
+/// `LeasedBatch` may only be dropped if it:
+/// - Does not have a leased `SeqNo (i.e. `self.leased_seqno.is_none()`)
+/// - Is consumed through `self.get_droppable_batch()`
+///
+/// In any other circumstance, dropping `LeasedBatch` panics.
+#[derive(Debug)]
+pub struct LeasedBatch<T>
+where
+    T: Timestamp + Codec64,
+{
+    pub(crate) shard_id: ShardId,
+    pub(crate) reader_id: ReaderId,
+    pub(crate) metadata: LeasedBatchMetadata<T>,
+    pub(crate) batch: HollowBatch<T>,
+    /// The `SeqNo` from which this batch originated; we track this value as
+    /// long as necessary to ensure the `SeqNo` isn't garbage collected while a
+    /// read still depends on it.
+    pub(crate) leased_seqno: Option<SeqNo>,
+    droppable_lease: bool,
+}
+
+impl<T> LeasedBatch<T>
+where
+    T: Timestamp + Codec64,
+{
+    pub(crate) fn new(
+        shard_id: ShardId,
+        reader_id: ReaderId,
+        metadata: LeasedBatchMetadata<T>,
+        batch: HollowBatch<T>,
+        leased_seqno: Option<SeqNo>,
+    ) -> LeasedBatch<T> {
+        LeasedBatch {
+            shard_id,
+            reader_id,
+            metadata,
+            batch,
+            leased_seqno,
+            droppable_lease: false,
+        }
+    }
+
+    /// Takes `self` into a [`SerdeLeasedBatch`], allowing users to drop `self`.
+    ///
+    /// !!!WARNING!!!
+    ///
+    /// If `self` has a `leased_seqno`, failing to take the returned
+    /// `SerdeLeasedBatch` back into a `LeasedBatch` will leak `SeqNo`s and
+    /// prevent persist compaction.
+    pub fn get_droppable_batch(mut self) -> SerdeLeasedBatch {
```

Comment:
as discussed offline, I feel we should unseparably tie together serialization and droppability. in my mental model, serializing a LeasedBatch _transfers_ the lease and allowed the LeasedBatch to be dropped. so, I think we can remove the droppable_lease field and instead clear the leased_seqno here (after the SerdeLeasedBatch::from call)